File: internal/capsule/wire.go
package capsule

import (
	"encoding/binary"
	"errors"
	"io"
	"net"
	"regexp"
	"unicode/utf8"

	"github.com/cyr1en/ref0/internal/safenet"
)

var operationID = regexp.MustCompile(`^[A-Za-z0-9._:-]{1,128}$`)

type wire struct {
	connection net.Conn
	limits     Limits
	aggregate  int
}

func newWire(connection net.Conn, limits Limits) *wire {
	return &wire{connection: connection, limits: limits}
}

func (transport *wire) send(message map[string]any) error {
	if err := validateWireMessage(message); err != nil {
		return err
	}
	encoded, err := canonicalJSON(message)
	if err != nil || len(encoded) == 0 || len(encoded) > transport.limits.MaxFrameBytes {
		return errors.New("capsule outbound frame exceeds limit")
	}
	if _, _, err := parseStrictJSON(encoded, complexityLimits{
		transport.limits.MaxDepth, transport.limits.MaxStringBytes, transport.limits.MaxKeys,
	}); err != nil {
		return errors.New("capsule outbound message exceeds JSON limits")
	}
	transport.aggregate += len(encoded) + 4
	if transport.aggregate > transport.limits.MaxAggregateBytes {
		return errors.New("capsule protocol aggregate exceeds limit")
	}
	frame := make([]byte, len(encoded)+4)
	binary.BigEndian.PutUint32(frame, uint32(len(encoded)))
	copy(frame[4:], encoded)
	return writeFull(transport.connection, frame)
}

func (transport *wire) receive() (map[string]any, error) {
	var header [4]byte
	if _, err := io.ReadFull(transport.connection, header[:]); err != nil {
		return nil, err
	}
	length := int(binary.BigEndian.Uint32(header[:]))
	if length < 1 || length > transport.limits.MaxFrameBytes {
		return nil, errors.New("capsule inbound frame exceeds limit")
	}
	transport.aggregate += length + 4
	if transport.aggregate > transport.limits.MaxAggregateBytes {
		return nil, errors.New("capsule protocol aggregate exceeds limit")
	}
	encoded := make([]byte, length)
	if _, err := io.ReadFull(transport.connection, encoded); err != nil {
		return nil, err
	}
	value, _, err := parseStrictJSON(encoded, complexityLimits{
		transport.limits.MaxDepth, transport.limits.MaxStringBytes, transport.limits.MaxKeys,
	})
	if err != nil {
		return nil, errors.New("capsule protocol JSON is invalid")
	}
	message, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("capsule protocol message must be an object")
	}
	if err := validateWireMessage(message); err != nil {
		return nil, err
	}
	return message, nil
}

func writeFull(writer io.Writer, value []byte) error {
	for len(value) != 0 {
		written, err := writer.Write(value)
		if err != nil {
			return err
		}
		if written == 0 {
			return io.ErrShortWrite
		}
		value = value[written:]
	}
	return nil
}

func validateWireMessage(message map[string]any) error {
	typeName, ok := message["type"].(string)
	if !ok {
		return errors.New("capsule protocol schema rejected a message")
	}
	var required, optional map[string]wireField
	switch typeName {
	case "start":
		required = map[string]wireField{
			"type": exactString("start"), "protocol_version": exactInteger(1), "attempt_id": operationIDField(),
			"role": enumString("PLANNER", "PAGE_WRITER"), "system_prompt": stringField(0, -1), "prompt": stringField(0, -1),
			"tools": toolArrayField(), "output_schema": objectField(), "provider": providerField(), "limits": limitsField(),
		}
	case "tool_result":
		required = map[string]wireField{
			"type": exactString("tool_result"), "id": operationIDField(), "result": anyField(), "content": stringField(0, -1),
		}
	case "model_result":
		required = map[string]wireField{
			"type": exactString("model_result"), "id": operationIDField(), "body_base64": stringField(0, -1),
		}
	case "cancel":
		required = map[string]wireField{"type": exactString("cancel"), "reason": stringField(0, 256)}
	case "tool_call":
		required = map[string]wireField{
			"type": exactString("tool_call"), "id": operationIDField(), "provider_call_id": operationIDField(),
			"name": stringField(1, 128), "arguments": objectField(),
		}
	case "model_request":
		required = map[string]wireField{
			"type": exactString("model_request"), "id": operationIDField(), "turn": integerRange(1, 1_000),
		}
	case "complete":
		required = map[string]wireField{"type": exactString("complete"), "output": objectField()}
	case "failed":
		required = map[string]wireField{
			"type": exactString("failed"), "code": enumString("cancelled", "protocol", "provider", "tool", "invalid_result", "internal"),
			"message": stringField(1, 400),
		}
	default:
		return errors.New("capsule protocol schema rejected a message")
	}
	for name, validator := range required {
		value, exists := message[name]
		if !exists || !validator(value) {
			return errors.New("capsule protocol schema rejected a message")
		}
	}
	for name, value := range message {
		validator, exists := required[name]
		if !exists {
			validator, exists = optional[name]
		}
		if !exists || !validator(value) {
			return errors.New("capsule protocol schema rejected a message")
		}
	}
	return nil
}

type wireField func(any) bool

func anyField() wireField { return func(any) bool { return true } }
func objectField() wireField {
	return func(value any) bool { _, ok := value.(map[string]any); return ok }
}
func exactString(want string) wireField {
	return func(value any) bool { item, ok := value.(string); return ok && item == want }
}
func exactInteger(want int64) wireField {
	return func(value any) bool { item, ok := jsonInteger(value); return ok && item == want }
}
func integerRange(low, high int64) wireField {
	return func(value any) bool { item, ok := jsonInteger(value); return ok && item >= low && item <= high }
}
func stringField(minimum, maximum int) wireField {
	return func(value any) bool {
		item, ok := value.(string)
		length := utf8.RuneCountInString(item)
		return ok && length >= minimum && (maximum < 0 || length <= maximum)
	}
}
func enumString(values ...string) wireField {
	allowed := set(values...)
	return func(value any) bool { item, ok := value.(string); _, exists := allowed[item]; return ok && exists }
}
func operationIDField() wireField {
	return func(value any) bool { item, ok := value.(string); return ok && validOperationID(item) }
}

func validOperationID(value string) bool { return operationID.MatchString(value) }

func toolArrayField() wireField {
	return func(value any) bool {
		items, ok := value.([]any)
		if !ok || len(items) > 64 {
			return false
		}
		for _, item := range items {
			tool, ok := item.(map[string]any)
			if !ok || len(tool) != 3 || !stringField(1, 128)(tool["name"]) || !stringField(0, 4_096)(tool["description"]) || !objectField()(tool["parameters"]) {
				return false
			}
			for name := range tool {
				if name != "name" && name != "description" && name != "parameters" {
					return false
				}
			}
		}
		return true
	}
}

func providerField() wireField {
	return func(value any) bool {
		provider, ok := value.(map[string]any)
		if !ok || len(provider) != 7 {
			return false
		}
		fields := map[string]wireField{
			"model_id": stringField(1, 512), "body_options": samplingField(),
			"reasoning_effort": enumString("none", "minimal", "low", "medium", "high", "max"),
			"context_window":   integerRange(1_024, 10_000_000), "max_output_tokens": integerRange(1, 1_000_000),
			"timeout_ms": integerRange(safenet.MinModelTimeout.Milliseconds(), safenet.MaxModelTimeout.Milliseconds()), "capsule_runtime_revision": stringField(1, 128),
		}
		for name, validator := range fields {
			if !validator(provider[name]) {
				return false
			}
		}
		for name := range provider {
			if _, exists := fields[name]; !exists {
				return false
			}
		}
		return true
	}
}

func samplingField() wireField {
	return func(value any) bool {
		object, ok := value.(map[string]any)
		if !ok {
			return false
		}
		_, err := validateSamplingOptions(object)
		return err == nil
	}
}

func limitsField() wireField {
	return func(value any) bool {
		object, ok := value.(map[string]any)
		if !ok || len(object) != 8 {
			return false
		}
		fields := map[string]wireField{
			"max_frame_bytes": integerRange(1_024, 4_194_304), "max_aggregate_bytes": integerRange(4_096, 33_554_432),
			"max_string_bytes": integerRange(256, 2_097_152), "max_depth": integerRange(4, 64),
			"max_keys": integerRange(32, 100_000), "max_fetch_body_bytes": integerRange(1_024, 1_048_576),
			"max_fetches": integerRange(1, 1_000), "max_model_requests": integerRange(1, 1_000),
		}
		for name, validator := range fields {
			if !validator(object[name]) {
				return false
			}
		}
		for name := range object {
			if _, exists := fields[name]; !exists {
				return false
			}
		}
		return true
	}
}

File: capsule/src/wire.ts
import { Type } from "typebox";
import type { Static } from "typebox";
import { Check } from "typebox/value";

import { checkCanonicalWireMessage } from "./canonical-schema.js";

const closed: { additionalProperties: false } = { additionalProperties: false };
const Id = Type.String({
  minLength: 1,
  maxLength: 128,
  pattern: "^[A-Za-z0-9._:-]+$",
});
const ReasoningEffort = Type.Union([
  Type.Literal("none"),
  Type.Literal("minimal"),
  Type.Literal("low"),
  Type.Literal("medium"),
  Type.Literal("high"),
  Type.Literal("max"),
]);
const Limits = Type.Object(
  {
    max_frame_bytes: Type.Integer({ minimum: 1024, maximum: 4_194_304 }),
    max_aggregate_bytes: Type.Integer({ minimum: 4096, maximum: 33_554_432 }),
    max_string_bytes: Type.Integer({ minimum: 256, maximum: 2_097_152 }),
    max_depth: Type.Integer({ minimum: 4, maximum: 64 }),
    max_keys: Type.Integer({ minimum: 32, maximum: 100_000 }),
    max_fetch_body_bytes: Type.Integer({ minimum: 1024, maximum: 1_048_576 }),
    max_fetches: Type.Integer({ minimum: 1, maximum: 1000 }),
    max_model_requests: Type.Integer({ minimum: 1, maximum: 1000 }),
  },
  closed,
);
const Provider = Type.Object(
  {
    model_id: Type.String({ minLength: 1, maxLength: 512 }),
    body_options: Type.Object(
      {
        temperature: Type.Optional(Type.Number({ minimum: 0, maximum: 2 })),
        top_p: Type.Optional(Type.Number({ minimum: 0, maximum: 1 })),
        frequency_penalty: Type.Optional(
          Type.Number({ minimum: -2, maximum: 2 }),
        ),
        presence_penalty: Type.Optional(
          Type.Number({ minimum: -2, maximum: 2 }),
        ),
        seed: Type.Optional(
          Type.Integer({ minimum: -2_147_483_648, maximum: 2_147_483_647 }),
        ),
        stop: Type.Optional(
          Type.Array(Type.String({ minLength: 1, maxLength: 1024 }), {
            minItems: 1,
            maxItems: 4,
          }),
        ),
      },
      closed,
    ),
    reasoning_effort: ReasoningEffort,
    context_window: Type.Integer({ minimum: 1024, maximum: 10_000_000 }),
    max_output_tokens: Type.Integer({ minimum: 1, maximum: 1_000_000 }),
    timeout_ms: Type.Integer({ minimum: 1_000, maximum: 60_000 }),
    capsule_runtime_revision: Type.String({ minLength: 1, maxLength: 128 }),
  },
  closed,
);
const Tool = Type.Object(
  {
    name: Type.String({ minLength: 1, maxLength: 128 }),
    description: Type.String({ maxLength: 4096 }),
    parameters: Type.Record(Type.String(), Type.Unknown()),
  },
  closed,
);
const Start = Type.Object(
  {
    type: Type.Literal("start"),
    protocol_version: Type.Literal(1),
    attempt_id: Id,
    role: Type.Union([Type.Literal("PLANNER"), Type.Literal("PAGE_WRITER")]),
    system_prompt: Type.String(),
    prompt: Type.String(),
    tools: Type.Array(Tool, { maxItems: 64 }),
    output_schema: Type.Record(Type.String(), Type.Unknown()),
    provider: Provider,
    limits: Limits,
  },
  closed,
);
const ToolResult = Type.Object(
  {
    type: Type.Literal("tool_result"),
    id: Id,
    result: Type.Unknown(),
    content: Type.String(),
  },
  closed,
);
const ModelResult = Type.Object(
  {
    type: Type.Literal("model_result"),
    id: Id,
    body_base64: Type.String(),
  },
  closed,
);
const Cancel = Type.Object(
  { type: Type.Literal("cancel"), reason: Type.String({ maxLength: 256 }) },
  closed,
);
const ToolCall = Type.Object(
  {
    type: Type.Literal("tool_call"),
    id: Id,
    provider_call_id: Id,
    name: Type.String({ minLength: 1, maxLength: 128 }),
    arguments: Type.Record(Type.String(), Type.Unknown()),
  },
  closed,
);
const ModelRequest = Type.Object(
  {
    type: Type.Literal("model_request"),
    id: Id,
    turn: Type.Integer({ minimum: 1, maximum: 1000 }),
  },
  closed,
);
const Complete = Type.Object(
  {
    type: Type.Literal("complete"),
    output: Type.Record(Type.String(), Type.Unknown()),
  },
  closed,
);
const Failed = Type.Object(
  {
    type: Type.Literal("failed"),
    code: Type.Union([
      Type.Literal("cancelled"),
      Type.Literal("protocol"),
      Type.Literal("provider"),
      Type.Literal("tool"),
      Type.Literal("invalid_result"),
      Type.Literal("internal"),
    ]),
    message: Type.String({ minLength: 1, maxLength: 400 }),
  },
  closed,
);

export const HostMessage = Type.Union([Start, ToolResult, ModelResult, Cancel]);
export const CapsuleMessage = Type.Union([
  ToolCall,
  ModelRequest,
  Complete,
  Failed,
]);
export type HostMessage = Static<typeof HostMessage>;
export type StartMessage = Static<typeof Start>;
export type CapsuleMessage = Static<typeof CapsuleMessage>;

export type ComplexityLimits = Readonly<{
  maxStringBytes: number;
  maxDepth: number;
  maxKeys: number;
}>;

export function validateComplexity(
  value: unknown,
  limits: ComplexityLimits,
): void {
  const pending: Array<readonly [unknown, number]> = [[value, 1]];
  let keys = 0;
  while (pending.length > 0) {
    const current = pending.pop();
    if (!current) break;
    const [item, depth] = current;
    if (depth > limits.maxDepth)
      throw new Error("protocol JSON depth exceeds limit");
    if (
      typeof item === "string" &&
      Buffer.byteLength(item) > limits.maxStringBytes
    ) {
      throw new Error("protocol JSON string exceeds limit");
    }
    if (Array.isArray(item)) {
      for (const child of item) pending.push([child, depth + 1]);
    } else if (typeof item === "object" && item !== null) {
      const entries = Object.entries(item);
      keys += entries.length;
      if (keys > limits.maxKeys)
        throw new Error("protocol JSON key count exceeds limit");
      for (const [key, child] of entries) {
        if (Buffer.byteLength(key) > limits.maxStringBytes)
          throw new Error("protocol JSON key exceeds limit");
        pending.push([child, depth + 1]);
      }
    }
  }
}

export function parseHostMessage(value: unknown): HostMessage {
  if (!checkCanonicalWireMessage(value) || !Check(HostMessage, value))
    throw new Error("invalid host protocol message");
  return value;
}

export function parseCapsuleMessage(value: unknown): CapsuleMessage {
  if (!checkCanonicalWireMessage(value) || !Check(CapsuleMessage, value))
    throw new Error("invalid capsule protocol message");
  return value;
}
